components/calendar/src/astronomy.rs

```diff
@@ -0,0 +1,657 @@
+            latitude: latitude_input,
+            longitude: longitude_input,
+            elevation: elevation_input,
+        };
+        result.check();
+        result
+    }
+
+    fn check(&self) {
+        if self.latitude < -90.0 || self.latitude > 90.0 {
+            debug_assert!(false, "Locations must have latitudes from -90.0 to 90.0.");
+        }
+        if self.longitude < -180.00 || self.longitude > 180.0 {
+            debug_assert!(false, "Locations must have longitudes from -180 to 180.");
+        }
+    }
+
+    /// Get the longitude of a location
+    ///
+    /// ```rust
+    /// use icu::calendar::astronomy::Location;
+    ///
+    /// let location: Location = Location::new(29.3, -132.6, 1032.5);
+    /// let longitude: f32 = location.longitude();
+    /// assert_eq!(longitude, -132.6);
+    /// ```
+    pub fn longitude(&self) -> f32 {
+        self.longitude
+    }
+
+    /// Get the latitude of a location
+    ///
+    /// ```rust
+    /// use icu::calendar::astronomy::Location;
+    ///
+    /// let location: Location = Location::new(29.3, -132.6, 1032.5);
+    /// let latitude: f32 = location.latitude();
+    /// assert_eq!(latitude, 29.3);
+    /// ```
+    pub fn latitude(&self) -> f32 {
+        self.latitude
+    }
+
+    /// Get the elevation of a location
+    ///
+    /// ```rust
+    /// use icu::calendar::astronomy::Location;
+    ///
+    /// let location: Location = Location::new(29.3, -132.6, 1032.5);
+    /// let elevation: f32 = location.elevation();
+    /// assert_eq!(elevation, 1032.5);
+    /// ```
+    pub fn elevation(&self) -> f32 {
+        self.elevation
+    }
+
+    /// Convert a longitude into a mean time zone;
+    /// this yields the difference in Moment given a longitude
+    /// e.g. a longitude of 90 degrees is 0.25 (90 / 360) days ahead
+    /// of a location with a longitude of 0 degrees.
+    pub fn zone_from_longitude(longitude: f64) -> f64 {
+        longitude / 360.0
+    }
+
+    /// Convert from local mean time to universal time given a location
+    pub fn universal_from_local(local_time: Moment, location: Location) -> Moment {
+        local_time - Self::zone_from_longitude(location.longitude as f64)
+    }
+
+    /// Convert from universal time to local time given a location
+    pub fn local_from_universal(universal_time: Moment, location: Location) -> Moment {
+        universal_time + Self::zone_from_longitude(location.longitude as f64)
+    }
+}
+
+#[derive(Debug)]
+pub struct Astronomical;
+
+impl Astronomical {
+    /// Function for the ephemeris correction, which corrects the
+    /// somewhat-unpredictable discrepancy between dynamical time
+    /// and universal time
+    ///
+    /// Reference code: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L3884-L3952
+    pub fn ephemeris_correction(moment: Moment) -> f64 {
+        let year = moment.inner().floor() / 365.2425;
+        let year_int = (if year > 0.0 { year + 1.0 } else { year }) as i32;
```

Comment:
Suggestion: the code to convert from Moment to an ISO year should be in its own function so it can be re-used and scrutinized